asab/storage/service.py

```diff
@@ -45,10 +63,105 @@ async def get(self, collection: str, obj_id) -> dict:
 
 
 	@abc.abstractmethod
-	async def get_by(self, collection: str, key: str, value):
+	async def get_by(self, collection: str, key: str, value, decrypt=None):
+		"""
+		Get object from collection by its key/value
+
+		:param collection: Collection to get from
+		:param key: Key to filter on
+		:param value: Value to filter on
+		:param decrypt: Set of fields to decrypt
+		:return: The object retrieved from a storage
+
+		Raises:
+			KeyError: If object {key: value} not found in `collection`
+		"""
 		pass
 
 
+	async def get_by_encrypted(self, collection: str, key: str, value, decrypt=None):
+		"""
+		Get object from collection by its key and encrypted value
+
+		:param collection: Collection to get from
+		:param key: Key to filter on
+		:param value: Value to encrypt and filter on
+		:param decrypt: Set of fields to decrypt
+		:return: The object retrieved from a storage
+
+		Raises:
+			KeyError: If object {key: value} not found in `collection`
+		"""
+		return await self.get_by(collection, key, self.aes_encrypt(value), decrypt)
+
+
 	@abc.abstractmethod
 	async def delete(self, collection: str, obj_id):
 		pass
+
+
+	def aes_encrypt(self, raw: bytes) -> bytes:
+		"""
+		Take a string or bytes and encrypt it using AES-CBC.
+
+		:param raw: The data to be encrypted
+		:type raw: bytes
+		:return: The encrypted data.
+		"""
+		if self.AESKey is None:
+			raise RuntimeError("No aes_key configured in asab:storage")
+
+		if not isinstance(raw, bytes):
+			if isinstance(raw, str):
+				raise TypeError("String objects must be encoded before encryption")
+			else:
+				raise TypeError("Only 'bytes' objects can be encrypted")
+
+		block_size = cryptography.hazmat.primitives.ciphers.algorithms.AES.block_size // 8
+		if len(raw) <= block_size:
+			raise ValueError("Cannot encrypt data shorter than {} bytes".format(block_size))
+
+		# Pad the text to fit the blocks
+		pad_length = -len(raw) % block_size
+		if pad_length != 0:
+			raw = raw + b"\00" * pad_length
+
+		# Keep the first block in plaintext, use it as initialization vector
+		iv, text_to_encrypt = raw[:block_size], raw[block_size:]
```

Comment:
This is problematic - the fist part of the `raw` is not encrypted ... I suggest to use explicit `iv` as a parameter here.

Reply:
Fixed.

If no `iv` is supplied, it generates a random one `using secrets.token_bytes`.